File: desktop-qt/shows/player.py
"""Thin wrapper over a python-mpv handle exposing the queue / wait /
clear / OSD operations the runner needs — Python analog of
desktop/internal/player. The mpv handle itself is owned by the QML
MpvItem; this wraps it for the runner thread.

mpv fires event callbacks on its own thread; we keep an end-file counter
under a condition variable that the runner thread blocks on.
"""

from __future__ import annotations

import threading
from typing import TYPE_CHECKING, Callable, Optional

if TYPE_CHECKING:
    # Only needed for the `mpv.MPV` annotation, which `from __future__ import
    # annotations` keeps as a string. Importing libmpv at module load would
    # otherwise force every importer (incl. the runner tests) to have the DLL.
    import mpv


class PlayerShutdown(Exception):
    """Raised when mpv emits a shutdown event (window closed)."""


class Player:
    def __init__(self, handle: mpv.MPV, on_pos: Optional[Callable[[int], None]] = None):
        self._m = handle
        self._cv = threading.Condition()
        self._end_files = 0
        self._shutdown = False

        @handle.event_callback("end-file")
        def _on_end(_ev):
            with self._cv:
                self._end_files += 1
                self._cv.notify_all()

        @handle.event_callback("shutdown")
        def _on_shutdown(_ev):
            with self._cv:
                self._shutdown = True
                self._cv.notify_all()

        # Fires when a queued file finishes loading — the runner uses it to seek
        # the now-playing episode to its saved resume position. Set after
        # construction (the runner is built later).
        self._on_file_loaded: Optional[Callable[[], None]] = None

        @handle.event_callback("file-loaded")
        def _on_loaded(_ev):
            cb = self._on_file_loaded
            if cb is not None:
                cb()

        # Keep refs so python-mpv's weakref-based callback registry doesn't
        # drop them.
        self._cbs = (_on_end, _on_shutdown, _on_loaded)

        # Report the current playlist index (which queued entry is playing) so
        # the overlay can show "now playing / up next" and the runner knows
        # which episode skip/defer act on. Fires on mpv's thread; the callback
        # must be thread-safe. Always registered; set_on_pos can (re)point it
        # after construction, since the runner that consumes it is built later.
        self._on_pos = on_pos

        def _pos_handler(_name, value):
            cb = self._on_pos
            if cb is not None and value is not None:
                cb(int(value))

        handle.observe_property("playlist-pos", _pos_handler)
        self._pos_handler = _pos_handler  # keep ref alive

    def set_on_pos(self, fn: Callable[[int], None]) -> None:
        """(Re)point the playlist-pos callback. Used by main.py to fan position
        updates out to both the overlay status and the runner once both exist."""
        self._on_pos = fn

    def set_on_file_loaded(self, fn: Callable[[], None]) -> None:
        self._on_file_loaded = fn

    def time_pos(self) -> Optional[float]:
        """Current playback position in seconds, or None if not playing."""
        try:
            return self._m.time_pos
        except Exception:
            return None

    def seek_absolute(self, seconds: float) -> None:
        """Seek to an absolute position in seconds — used to restore a resume
        point when an episode loads."""
        try:
            self._m.command("seek", seconds, "absolute")
        except Exception:
            pass

    # ── commands ──────────────────────────────────────────────────────
    def play(self, path: str, mode: str) -> None:
        # mode: "replace" for the first entry, "append-play" for the rest.
        self._m.loadfile(path, mode)

    def playlist_clear(self) -> None:
        try:
            self._m.command("playlist-clear")
        except Exception:
            pass

    def show_text(self, text: str, duration_ms: int) -> None:
        try:
            self._m.command("show-text", text, str(duration_ms))
        except Exception:
            pass

    def set_pause(self, paused: bool) -> None:
        self._m.pause = paused

    def toggle_pause(self) -> None:
        self._m.pause = not self._m.pause

    def skip(self) -> None:
        # Force-advance to the next queued entry (the runner counts the
        # resulting end-file like a natural one).
        try:
            self._m.command("playlist-next", "force")
        except Exception:
            pass

    # ── playback controls (seek / volume / tracks) ─────────────────────
    def seek_relative(self, seconds: float) -> None:
        """Seek by `seconds` from the current position (negative = back)."""
        try:
            self._m.command("seek", seconds, "relative")
        except Exception:
            pass

    def seek_percent(self, pct: float) -> None:
        """Seek to an absolute percentage (0-100) of the file — what the
        scrub bar posts."""
        try:
            self._m.command("seek", max(0.0, min(100.0, pct)), "absolute-percent")
        except Exception:
            pass

    def set_volume(self, vol: float) -> None:
        try:
            self._m.volume = max(0.0, min(130.0, vol))
        except Exception:
            pass

    def set_sub(self, sid) -> None:
        """sid is an int track id, or "no" to disable subtitles."""
        try:
            self._m.sid = sid
        except Exception:
            pass

    def set_audio(self, aid) -> None:
        try:
            self._m.aid = aid
        except Exception:
            pass

    def playback_state(self) -> dict:
        """Snapshot of current playback for the overlay's scrub bar, volume,
        and track menus. Reads mpv properties; safe from the HTTP thread."""
        def g(name, default=None):
            try:
                return getattr(self._m, name)
            except Exception:
                return default

        def tracks(kind):
            out = []
            for t in (g("track_list", []) or []):
                if t.get("type") != kind:
                    continue
                tid = t.get("id")
                out.append({
                    "id": tid,
                    "title": t.get("title") or t.get("lang") or f"{kind} {tid}",
                    "lang": t.get("lang"),
                    "selected": bool(t.get("selected")),
                })
            return out

        return {
            "time_pos": g("time_pos"),
            "duration": g("duration"),
            "percent_pos": g("percent_pos"),
            "volume": g("volume"),
            "paused": bool(g("pause", False)),
            "sub_tracks": tracks("sub"),
            "audio_tracks": tracks("audio"),
            "sid": g("sid"),
            "aid": g("aid"),
        }

    # ── round synchronization ─────────────────────────────────────────
    def wait_for_round(self, n: int, stop: threading.Event) -> None:
        """Block until `n` more end-file events arrive. Raises
        PlayerShutdown if mpv closes; returns early (raising) if stopped."""
        with self._cv:
            target = self._end_files + n
            while self._end_files < target:
                if self._shutdown:
                    raise PlayerShutdown("mpv shutdown event")
                if stop.is_set():
                    raise PlayerShutdown("runner stopped")
                self._cv.wait(timeout=0.5)

File: desktop-rs/desktop/src/player.rs
//! The libmpv player: wraps an [`mpv::Handle`] with the queue/wait/clear/OSD
//! operations the runner needs (via [`shows_core::runner::PlayerOps`]) and the
//! seek/volume/track controls the control server needs. An event-pump thread
//! turns libmpv's C events into the runner's callbacks.

use std::ffi::{CStr, c_char, c_int, c_void};
use std::sync::{Arc, Condvar, Mutex};
use std::time::Duration;

use serde_json::{Value, json};

use shows_core::runner::{PlayerOps, PlayerShutdown, RoundEndTracker, StopFlag};

use crate::mpv::{
    Handle, MPV_END_FILE_REASON_EOF, MPV_EVENT_END_FILE, MPV_EVENT_FILE_LOADED, MPV_EVENT_SHUTDOWN,
    MpvEventEndFile,
};

const MPV_EVENT_PROPERTY_CHANGE: c_int = 22;
const MPV_FORMAT_STRING: c_int = 1;

#[repr(C)]
struct MpvEventProperty {
    name: *const c_char,
    format: c_int,
    data: *mut c_void,
}

type Cb = Box<dyn Fn() + Send + Sync>;
type PosCb = Box<dyn Fn(usize) + Send + Sync>;

const PLAYBACK_OBSERVED_PROPERTIES: &[&str] = &[
    "time-pos",
    "duration",
    "percent-pos",
    "volume",
    "pause",
    "sid",
    "aid",
    "track-list/count",
];

struct Shared {
    // Round boundary, driven by playlist-pos (-1 == exhausted). The tracker
    // counts a round end once per real exhaustion and ignores the startup/idle
    // -1, so in-round navigation (skip/prev) never miscounts the boundary.
    round: RoundEndTracker,
    shutdown: bool,
    interrupted: bool,
}

pub struct Player {
    handle: Arc<Handle>,
    cv: Condvar,
    shared: Mutex<Shared>,
    on_natural_end: Mutex<Option<Cb>>,
    on_file_loaded: Mutex<Option<Cb>>,
    on_playback_change: Mutex<Option<Cb>>,
    on_pos: Mutex<Option<PosCb>>,
}

impl Player {
    /// Wrap the handle and start the event pump. The handle is shared with the
    /// compositor's render context, so it's an `Arc`.
    pub fn new(handle: Arc<Handle>) -> Arc<Player> {
        let player = Arc::new(Player {
            handle,
            cv: Condvar::new(),
            shared: Mutex::new(Shared {
                round: RoundEndTracker::default(),
                shutdown: false,
                interrupted: false,
            }),
            on_natural_end: Mutex::new(None),
            on_file_loaded: Mutex::new(None),
            on_playback_change: Mutex::new(None),
            on_pos: Mutex::new(None),
        });
        // Report which queued entry is playing so the runner routes skip/defer to
        // it and the overlay shows now-playing.
        player.handle.observe_property_string(1, "playlist-pos");
        for (i, name) in PLAYBACK_OBSERVED_PROPERTIES.iter().enumerate() {
            player.handle.observe_property_string(10 + i as u64, name);
        }
        let pump = player.clone();
        std::thread::Builder::new()
            .name("mpv-events".into())
            .spawn(move || pump.event_loop())
            .expect("spawn mpv event pump");
        player
    }

    pub fn set_on_natural_end(&self, f: Cb) {
        *self.on_natural_end.lock().unwrap() = Some(f);
    }
    pub fn set_on_file_loaded(&self, f: Cb) {
        *self.on_file_loaded.lock().unwrap() = Some(f);
    }
    pub fn set_on_playback_change(&self, f: Cb) {
        *self.on_playback_change.lock().unwrap() = Some(f);
    }
    pub fn set_on_pos(&self, f: PosCb) {
        *self.on_pos.lock().unwrap() = Some(f);
    }

    fn notify_playback_change(&self) {
        if let Some(cb) = self.on_playback_change.lock().unwrap().as_ref() {
            cb();
        }
    }

    fn event_loop(self: Arc<Self>) {
        loop {
            let ev = self.handle.wait_event(-1.0); // block until the next event
            if ev.is_null() {
                continue;
            }
            let event_id = unsafe { (*ev).event_id };
            match event_id {
                MPV_EVENT_SHUTDOWN => {
                    {
                        let mut s = self.shared.lock().unwrap();
                        s.shutdown = true;
                    }
                    self.cv.notify_all();
                    return; // mpv core is going away; stop pumping
                }
                MPV_EVENT_END_FILE => {
                    // Only a natural end (played to completion) marks the episode
                    // watched. The round boundary is detected from playlist-pos
                    // -> -1 (below), not by counting end-file events, so a
                    // replayed entry (playlist-prev) never miscounts the round.
                    let natural = unsafe {
                        let d = (*ev).data as *const MpvEventEndFile;
                        !d.is_null() && (*d).reason == MPV_END_FILE_REASON_EOF
                    };
                    if natural {
                        if let Some(cb) = self.on_natural_end.lock().unwrap().as_ref() {
                            cb();
                        }
                    }
                }
                MPV_EVENT_FILE_LOADED => {
                    if let Some(cb) = self.on_file_loaded.lock().unwrap().as_ref() {
                        cb();
                    }
                    self.notify_playback_change();
                }
                MPV_EVENT_PROPERTY_CHANGE => unsafe {
                    let prop = (*ev).data as *const MpvEventProperty;
                    if prop.is_null() {
                        continue;
                    }
                    let name = if (*prop).name.is_null() {
                        ""
                    } else {
                        CStr::from_ptr((*prop).name).to_str().unwrap_or("")
                    };
                    if name == "playlist-pos"
                        && (*prop).format == MPV_FORMAT_STRING
                        && !(*prop).data.is_null()
                    {
                        let sptr = *((*prop).data as *const *const c_char);
                        if !sptr.is_null() {
                            if let Ok(i) = CStr::from_ptr(sptr).to_string_lossy().parse::<i64>() {
                                // Feed the tracker: a -1 after real playback is
                                // the round boundary that wakes
                                // wait_for_round_end (keep-open=no lets the last
                                // entry advance past its end so this fires
                                // instead of pausing on the final frame). The
                                // startup/idle -1 returns false and is ignored.
                                let ended = self.shared.lock().unwrap().round.observe(i);
                                if ended {
                                    self.cv.notify_all();
                                }
                                if i >= 0 {
                                    if let Some(cb) = self.on_pos.lock().unwrap().as_ref() {
                                        cb(i as usize);
                                    }
                                }
                            }
                        }
                    }
                    if PLAYBACK_OBSERVED_PROPERTIES.contains(&name) {
                        self.notify_playback_change();
                    }
                },
                _ => {}
            }
        }
    }

    // ── control-server playback controls ─────────────────────────────────
    pub fn toggle_pause(&self) {
        self.handle.command(&["cycle", "pause"]);
    }
    pub fn seek_relative(&self, seconds: f64) {
        self.handle
            .command(&["seek", &seconds.to_string(), "relative"]);
    }
    pub fn seek_percent(&self, pct: f64) {
        self.handle.command(&[
            "seek",
            &pct.clamp(0.0, 100.0).to_string(),
            "absolute-percent",
        ]);
    }
    pub fn set_volume(&self, vol: f64) {
        self.handle
            .set_property("volume", &vol.clamp(0.0, 130.0).to_string());
    }
    pub fn set_sub(&self, sid: &str) {
        self.handle.set_property("sid", sid); // an id, or "no" to disable
    }
    pub fn set_audio(&self, aid: &str) {
        self.handle.set_property("aid", aid);
    }

    fn prop_f64(&self, name: &str) -> Option<f64> {
        self.handle
            .get_property(name)
            .and_then(|s| s.parse::<f64>().ok())
    }

    fn tracks(&self, kind: &str) -> Vec<Value> {
        let count: i64 = self
            .handle
            .get_property("track-list/count")
            .and_then(|s| s.parse().ok())
            .unwrap_or(0);
        let mut out = Vec::new();
        for i in 0..count {
            if self
                .handle
                .get_property(&format!("track-list/{i}/type"))
                .as_deref()
                != Some(kind)
            {
                continue;
            }
            let id = self
                .handle
                .get_property(&format!("track-list/{i}/id"))
                .unwrap_or_default();
            let title = self.handle.get_property(&format!("track-list/{i}/title"));
            let lang = self.handle.get_property(&format!("track-list/{i}/lang"));
            let selected = self
                .handle
                .get_property(&format!("track-list/{i}/selected"))
                .map(|s| s == "yes")
                .unwrap_or(false);
            out.push(json!({
                "id": id.parse::<i64>().ok(),
                "title": title.clone().or_else(|| lang.clone()).unwrap_or_else(|| format!("{kind} {id}")),
                "lang": lang,
                "selected": selected,
            }));
        }
        out
    }

    /// Snapshot of current playback for the overlay's scrub bar + track menus.
    pub fn playback_state(&self) -> Value {
        json!({
            "time_pos": self.prop_f64("time-pos"),
            "duration": self.prop_f64("duration"),
            "percent_pos": self.prop_f64("percent-pos"),
            "volume": self.prop_f64("volume"),
            "paused": self.handle.get_property("pause").map(|s| s == "yes").unwrap_or(false),
            "sub_tracks": self.tracks("sub"),
            "audio_tracks": self.tracks("audio"),
            "sid": self.handle.get_property("sid"),
            "aid": self.handle.get_property("aid"),
        })
    }

    /// Retrieve the width and height of the currently playing video.
    pub fn video_dimensions(&self) -> Option<(i32, i32)> {
        let w = self.prop_f64("video-params/w")? as i32;
        let h = self.prop_f64("video-params/h")? as i32;
        if w > 0 && h > 0 { Some((w, h)) } else { None }
    }
}

impl PlayerOps for Player {
    fn play(&self, path: &str, mode: &str) {
        // mode: "replace" for the first entry, "append-play" for the rest.
        self.handle.command(&["loadfile", path, mode]);
    }
    fn playlist_clear(&self) {
        {
            let mut s = self.shared.lock().unwrap();
            s.interrupted = true;
        }
        self.cv.notify_all();
        self.handle.command(&["playlist-clear"]);
    }
    fn show_text(&self, text: &str, duration_ms: i64) {
        self.handle
            .command(&["show-text", text, &duration_ms.to_string()]);
    }
    fn skip(&self) {
        // Force-advance to the next queued entry. At the last entry this
        // exhausts the playlist (playlist-pos -> -1), which ends the round.
        self.handle.command(&["playlist-next", "force"]);
    }
    fn previous(&self) {
        // Step back to the previous queued entry. "weak" so going back at the
        // first entry is a no-op rather than terminating playback.
        self.handle.command(&["playlist-prev", "weak"]);
    }
    fn time_pos(&self) -> Option<f64> {
        self.prop_f64("time-pos")
    }
    fn seek_absolute(&self, seconds: f64) {
        self.handle
            .command(&["seek", &seconds.to_string(), "absolute"]);
    }
    fn set_playlist_pos(&self, idx: usize) {
        self.handle.set_property("playlist-pos", &idx.to_string());
    }
    fn wait_for_round_end(&self, stop: &StopFlag) -> Result<(), PlayerShutdown> {
        let mut s = self.shared.lock().unwrap();
        s.interrupted = false; // Reset interrupted status before waiting
        let target = s.round.ended_count() + 1;
        while s.round.ended_count() < target {
            if s.shutdown {
                return Err(PlayerShutdown("mpv shutdown event".into()));
            }
            if stop.is_set() {
                return Err(PlayerShutdown("runner stopped".into()));
            }
            if s.interrupted {
                return Err(PlayerShutdown("interrupted".into()));
            }
            let (guard, _) = self.cv.wait_timeout(s, Duration::from_millis(500)).unwrap();
            s = guard;
        }
        Ok(())
    }
}
